File: TTdsa-java/Exercise/src/com/aacirq/string/N44.java
package com.aacirq.string;

public class N44 {
    public static void main(String[] args) {
        N44 sol = new N44();

        // "I am a nowcoder."
        System.out.println("\"" + sol.ReverseSentence("nowcoder. a am I") + "\"");
        // ""
        System.out.println("\"" + sol.ReverseSentence("") + "\"");
        // " "
        System.out.println("\"" + sol.ReverseSentence(" ") + "\"");
        // "  "
        System.out.println("\"" + sol.ReverseSentence("  ") + "\"");
    }

    public String ReverseSentence(String str) {
        if (str.length() == 0)
            return str;
        StringBuilder res = new StringBuilder();
        int sInd = str.length() - 1;
        int eInd = sInd;
        while (sInd >= 0) {
            while (sInd >= 0 && str.charAt(sInd) != ' ')
                sInd--;
            res.append(str.substring(sInd + 1, eInd + 1));
            if (sInd >= 0)
                res.append(' ');
            eInd = sInd - 1;
            sInd = eInd;
        }
        return res.toString();
    }
}

File: TTdsa-cpp/Exercise/string/N44.cpp
//
// Created by renxin on 2021/6/11.
//

#include "ttdsa.h"

class Solution {
public:
    string ReverseSentence(string str) {
        if (str.size() == 0)
            return str;
        string res(str.size(), ' ');
        int ind = 0;
        int eInd = str.size() - 1;
        int sInd = eInd;
        while (ind < res.size()) {
            while (sInd >= 0 && str[sInd] != ' ')
                --sInd;
            for (int i = sInd + 1; i <= eInd; i++) {
                res[ind++] = str[i];
            }
            ind++;
            eInd = sInd - 1;
            sInd = eInd;
        }
        return res;
    }
};

int main() {
    Solution sol;

    // "I am a nowcoder."
    cout << "\"" << sol.ReverseSentence("nowcoder. a am I") << "\"" << endl;
    // ""
    cout << "\"" << sol.ReverseSentence("") << "\"" << endl;
    // " "
    cout << "\"" << sol.ReverseSentence(" ") << "\"" << endl;
    // "  "
    cout << "\"" << sol.ReverseSentence("  ") << "\"" << endl;
    return 0;
}